armsrc/mifaresniff.c

```diff
@@ -1,177 +1,184 @@
+				sniffSAK = data[0];
+				if ((sniffUID[3] == 0x88) && (sniffUIDType == SNF_UID_4)) {			// CL2 UID part to be expected
+					sniffUIDType = SNF_UID_7;
+					memcpy(sniffUID, sniffUID + 4, 3);
+					sniffState = SNF_UID2;
+				} else {															// select completed
+					sniffState = SNF_CARD_IDLE;
+				}
+			}
+			break;
+		}
+		case SNF_ANTICOL2:{
+			if ((!reader) && (len == 5) && ((data[0] ^ data[1] ^ data[2] ^ data[3]) == data[4])) { // CL2 UID 
+				memcpy(sniffUID + 3, data, 4);
+				sniffState = SNF_UID2;
+			}
+			break;
+		}
 		case SNF_UID2:{
-			if ((reader) && (len == 9) && (data[0] == 0x95) && (data[1] == 0x70) && (CheckCrc14443(CRC_14443_A, data, 9))) {	// Select 2nd part of 7 Byte UID
+			if ((reader) && (len == 2) && (data[0] == 0x95) && (data[1] == 0x20)) {
+				sniffState = SNF_ANTICOL2;
+			}
+			
+			if ((reader) && (len == 9) && (data[0] == 0x95) && (data[1] == 0x70) && (CheckCrc14443(CRC_14443_A, data, 9))) {
+				memcpy(sniffUID + 3, &data[2], 4);
 				sniffState = SNF_SAK;
 			}
 			break;
 		}
-		case SNF_CARD_IDLE:{	// trace the card select sequence
-			sniffBuf[0] = 0xFF;
-			sniffBuf[1] = 0xFF;
-			memcpy(sniffBuf + 2, sniffUID, 7);
-			memcpy(sniffBuf + 9, sniffATQA, 2);
-			sniffBuf[11] = sniffSAK;
-			sniffBuf[12] = 0xFF;
-			sniffBuf[13] = 0xFF;
-			LogTrace(sniffBuf, 14, 0, 0, NULL, TRUE);
-			sniffState = SNF_CARD_CMD;
-		}	// intentionally no break;
-		case SNF_CARD_CMD:{	
-			LogTrace(data, len, 0, 0, NULL, reader);
-			timerData = GetTickCount();
-			break;
-		}
-	
-		default:
-			sniffState = SNF_INIT;
-		break;
-	}
-
-
-	return FALSE;
-}
-
-bool RAMFUNC MfSniffSend(uint16_t maxTimeoutMs) {
-	if (BigBuf_get_traceLen() && (GetTickCount() > timerData + maxTimeoutMs)) {
-		return intMfSniffSend();
-	}
-	return FALSE;
-}
-
-// internal sending function. not a RAMFUNC.
-bool intMfSniffSend() {
-
-	int pckSize = 0;
-	int pckLen = BigBuf_get_traceLen();
-	int pckNum = 0;
-	uint8_t *trace = BigBuf_get_addr();
-	
-	FpgaDisableSscDma();
-	while (pckLen > 0) {
-		pckSize = MIN(USB_CMD_DATA_SIZE, pckLen);
-		LED_B_ON();
-		cmd_send(CMD_ACK, 1, BigBuf_get_traceLen(), pckSize, trace + BigBuf_get_traceLen() - pckLen, pckSize);
-		LED_B_OFF();
-
-		pckLen -= pckSize;
-		pckNum++;
-	}
-
-	LED_B_ON();
-	cmd_send(CMD_ACK,2,0,0,0,0);
-	LED_B_OFF();
-
-	clear_trace();
-	
-	return TRUE;
-}
+		case SNF_CARD_IDLE:{	// trace the card select sequence
+			sniffBuf[0] = 0xFF;
+			sniffBuf[1] = 0xFF;
+			memcpy(sniffBuf + 2, sniffUID, 7);
+			memcpy(sniffBuf + 9, sniffATQA, 2);
+			sniffBuf[11] = sniffSAK;
+			sniffBuf[12] = 0xFF;
+			sniffBuf[13] = 0xFF;
+			LogTrace(sniffBuf, 14, 0, 0, NULL, TRUE);
+			sniffState = SNF_CARD_CMD;
+		}	// intentionally no break;
+		case SNF_CARD_CMD:{	
+			LogTrace(data, len, 0, 0, NULL, reader);
+			timerData = GetTickCount();
+			break;
+		}
+	
+		default:
+			sniffState = SNF_INIT;
+		break;
+	}
+
+
+	return FALSE;
+}
+
+bool RAMFUNC MfSniffSend(uint16_t maxTimeoutMs) {
```

Comment:
This may have been a good idea as workaround when BigBuf was limited to a few kilobytes. However, this has always been dangerous because the reader may again start sending commands (which we will miss) while we are sending the trace to client.

I propose to continue tracing instead until button or key pressed (like we do with hf 14a snoop).

Reply:
I think about send data without buffer clearing, but it needs to realize if it possible. so i will return to this after emv emul.